apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java

```diff
@@ -183,8 +183,16 @@ protected void bindErrorHandlers() {
       bind(HttpErrorHandler.class).annotatedWith(ServerError.class).to(ParseS3ErrorFromXmlContent.class);
    }
 
-   protected void bindRequestSigner() {
-      bind(RequestAuthorizeSignature.class).to(RequestAuthorizeSignatureV2.class).in(Scopes.SINGLETON);
+   @Provides
+   @Singleton
+   protected void bindRequestSigner(@Named(Constants.PROPERTY_V4_REQUEST_SIGNATURES) boolean v4Signatures) {
+      Class<? extends RequestAuthorizeSignature> clazz;
+      if (v4Signatures) {
+         clazz = RequestAuthorizeSignatureV4.class;
+      } else {
+         clazz = RequestAuthorizeSignatureV2.class;
+      }
+      bind(RequestAuthorizeSignature.class).to(clazz).in(Scopes.SINGLETON);
```

Comment:
Hmm, does this actually work? I don't know if `bind` statements can fo in a provider method, but it looks semantically weird. Provider methods are supposed to return objects that can be injected. Instead, I'd change this method to something like this (I haven't tried it, but you'll get the idea):

```java
@Provides
@Singleton
protected RequestAuthorizeSignature bindRequestSigner(
      @Named(Constants.PROPERTY_V4_REQUEST_SIGNATURES) boolean v4Signatures,
      RequestAuthorizeSignatureV2 v2,
      RequestAuthorizeSignatureV4 v4) {
   return v4Signatures ? v4 : v2;
}
```

The idea is to have the provider method get injected the possible instances and return the right one when an injection is requested.

Adding a test for this should be pretty straightforward. I'd just create a test class that creates a raw Guice injector with this module and the value of the property, then call `injector.getInstance` and verify that the returned object is of the expected type.

Reply:
A better approach to avoid instantiating the objects you're not using would be something like:

```java
@Provides
@Singleton
protected RequestAuthorizeSignature bindRequestSigner(
      @Named(Constants.PROPERTY_V4_REQUEST_SIGNATURES) boolean v4Signatures,
      Injector i) {
   return i.getInstance(v4Signatures ? RequestAuthorizeSignatureV4.class : RequestAuthorizeSignatureV2.class);
}
```

Reply:
Thanks for the review @nacx . You're right about the `bind` in a provider method - I have not run this yet. Will do so in a unit test.

Reply:
> Adding a test for this should be pretty straightforward. I'd just create a test class that creates a raw Guice injector with this module and the value of the property, then call `injector.getInstance` and verify that the returned object is of the expected type.

Is there a recommended way to handle the missing dependencies of this module in the test? I gather that this module requires quite a few dependencies from other modules which do not have mocks. 

I suppose I could extract the provider of the `RequestAuthorizeSignature` to its own module, that way its testing should be simpler.